pkg/execution/state/redis_state/backlog.go

```diff
@@ -0,0 +1,284 @@
+
+			// This is always specified, even if no key was configured in the function definition.
+			// In that case, the Throttle Key is the hashed function ID. See Schedule() for more details.
+			ThrottleKey: &i.Data.Throttle.Key,
+		}
+
+		if i.Data.Throttle.UnhashedThrottleKey != "" {
+			unhashedKey := i.Data.Throttle.UnhashedThrottleKey
+			// truncate - just in case
+			if len(unhashedKey) > 512 {
+				unhashedKey = unhashedKey[:512]
+			}
+			b.ThrottleKeyRawValue = &unhashedKey
+		}
+
+		backlogs = append(backlogs, b)
+	}
+
+	concurrencyKeys := i.Data.GetConcurrencyKeys()
+
+	// NOTE: This is an optimization that ensures we return *updated* concurrency keys
+	// for any recently published function configuration.  The embeddeed ckeys from the
+	// queue items above may be outdated.
+	if q.customConcurrencyLimitRefresher != nil {
+		// As an optimization, allow fetching updated concurrency limits if desired.
+		updated, _ := duration(ctx, q.primaryQueueShard.Name, "backlog_custom_concurrency_refresher", q.clock.Now(), func(ctx context.Context) ([]state.CustomConcurrency, error) {
+			return q.customConcurrencyLimitRefresher(ctx, i), nil
+		})
+		for _, update := range updated {
+			// This is quadratic, but concurrency keys are limited to 2 so it's
+			// okay.
+			for n, existing := range concurrencyKeys {
+				if existing.Key == update.Key {
+					concurrencyKeys[n].Limit = update.Limit
+				}
+			}
+		}
+	}
+
+	// Create concurrency key backlogs
+	for _, key := range concurrencyKeys {
+		scope, _, checksum, _ := key.ParseKey()
+
+		var rawValue *string
+		if key.UnhashedEvaluatedKeyValue != "" {
+			rawValue = &key.UnhashedEvaluatedKeyValue
+		}
+
+		backlogs = append(backlogs, QueueBacklog{
+			BacklogID:        fmt.Sprintf("conc:%s", key.Key),
+			ConcurrencyScope: &scope,
+
+			// Hashed expression to identify which key this is in the shadow partition concurrency key list
+			ConcurrencyKey: &key.Hash,
+
+			// Evaluated hashed and unhashed values
+			ConcurrencyKeyValue:         &checksum,
+			ConcurrencyKeyUnhashedValue: rawValue,
+		})
+	}
+
+	// Use default backlog if no concurrency/throttle backlogs are set up
+	if len(backlogs) == 0 {
+		backlogs = append(backlogs, QueueBacklog{
+			BacklogID: fmt.Sprintf("default:%s", i.FunctionID),
+		})
+	}
+
+	return backlogs
+}
+
+func (q *queue) ItemShadowPartition(ctx context.Context, i osqueue.QueueItem) QueueShadowPartition {
+	var (
+		ckeys = i.Data.GetConcurrencyKeys()
+	)
+
+	queueName := i.QueueName
+
+	// sanity check: both QueueNames should be set, but sometimes aren't
+	if queueName == nil && i.QueueName != nil {
+		queueName = i.QueueName
+		q.logger.Warn().Interface("item", i).Msg("shadow partitions encountered queue item with inconsistent custom queue name, should have both i.QueueName and i.Data.QueueName set")
+	}
+
+	// sanity check: queueName values must match
+	if i.Data.QueueName != nil && i.QueueName != nil && *i.Data.QueueName != *i.QueueName {
```

Comment:
this is too hard to read, wrap it in a function that's more descriptive.

```
i.hasDifferentQueueName()
```
or something